site/src/pages/WorkspacesPage/data.ts

```diff
@@ -0,0 +1,113 @@
+import {
+  QueryKey,
+  useMutation,
+  useQuery,
+  useQueryClient,
+} from "@tanstack/react-query"
+import { getWorkspaces, updateWorkspaceVersion } from "api/api"
+import {
+  Workspace,
+  WorkspaceBuild,
+  WorkspacesResponse,
+} from "api/typesGenerated"
+
+const getQueryKey = (page: number, query: string) => {
+  return ["workspaces", query, page]
+}
+
+type UseWorkspacesDataParams = {
+  page: number
+  limit: number
+  query: string
+}
+
+export const useWorkspacesData = ({
+  page,
+  limit,
+  query,
+}: UseWorkspacesDataParams) => {
+  const queryKey = getQueryKey(page, query)
+  const result = useQuery({
+    queryKey,
+    queryFn: () =>
+      getWorkspaces({
+        q: query,
+        limit: limit,
+        offset: page,
+      }),
+    refetchInterval: 5_000,
+  })
+
+  return {
+    ...result,
+    queryKey,
+  }
+}
+
+export const useWorkspaceUpdate = (queryKey: QueryKey) => {
+  const queryClient = useQueryClient()
+
+  return useMutation({
+    mutationFn: updateWorkspaceVersion,
+    onMutate: async (workspace) => {
+      await queryClient.cancelQueries({ queryKey })
+      queryClient.setQueryData<WorkspacesResponse>(queryKey, (oldResponse) => {
+        if (oldResponse) {
+          return assignPendingStatus(oldResponse, workspace)
+        }
+      })
+    },
+    onSuccess: (workspaceBuild) => {
+      queryClient.setQueryData<WorkspacesResponse>(queryKey, (oldResponse) => {
```

Comment:
Why would the response on `onSuccess` be `oldResponse`? It's not an updated `WorkspacesResponse`?

Reply:
The `updateWorkspaceVersion` function does not return a `WorkspacesResponse`. It returns a workspace build of an individual workspace, so what we have to do is get the old response, find the workspace that was updated, and replace the last build with the one returned from the `updateWorkspaceVersion` response.